operator: allow configurable K8s client rate limit with higher defaults

thorn3r opened the pull request:
The rate limit for K8s client(s) for Cilium Operator can now be
configured separately from that of Cilium Agent with the new
`--operator-k8s-client-qps` and `--operator-k8s-client-burst` flags.

To facilitate sharing a set of default configuration options for all
clients, the `Config` struct was split into `SharedConfig` and
`ClientParams`, the latter of which contains options expected to differ.

With K8s 1.26 reaching EOL, the default rate limit for Cilium Agent K8s
clients is now determined by the default config, rather than relying on
the upstream client library.

Since the Operator is a Deployment and does not scale linearly with
cluster size like the Agent DaemonSet, the rate limit can be increased
to improve performance at scale without heavily impacting control plane
resources. The default rate limit for Operator clients has been increased
to 100 QPS and 200 Burst.

```release-note
Add options to configure rate limit of Cilium Operator's K8s client.
Increase default rate limit of Cilium Operator's K8s client to 100 QPS.
```

pippolo84 commented on pkg/k8s/client/cell.go:
```diff
@@ -55,12 +55,13 @@ var Cell = cell.Module(
+// client.OperatorClientBuilderCell provides a function to create a new composite Clientset,
 // allowing a controller to use its own Clientset with a different user agent.
-var ClientBuilderCell = cell.Module(
+var OperatorClientBuilderCell = cell.Module(
```
Since this is specific to the operator, why not move it under `operator/` (something like `operator/pkg/k8s/client` or `operator/k8s/client`)?
We can do that for the operator clientset cell, def config, `ClientBuilderFunc` type, `NewClientBuilder` and everything that is operator-specific, leaving here the agent cell and the common functions.

pippolo84 commented on operator/cmd/root.go:
```diff
@@ -98,11 +98,12 @@ var (
+
+		// Provides a Clientset for accessing the Kubernetes API using the provided ClientBuilderFunc.
+		// This client is shared by controllers that do not create their own via the ClientBuilderFunc.
+		cell.Provide(func(f k8sClient.ClientBuilderFunc) (k8sClient.Clientset, error) { return f("") }),
```
Nit: we can group these two with:

```go
var OperatorK8sClientCell = cell.Group(
	k8sClient.OperatorClientBuilderCell,
	cell.Provide(func(f k8sClient.ClientBuilderFunc) (k8sClient.Clientset, error) { return f("") }),
),
```

and move this too under `operator/` (see the other comment). Then, we just need to add `OperatorK8sClientCell` here.

Also, I wonder if it might be useful to specify "cilium-operator" for the client name. It should help differentiate between this client requests and the agents' ones when looking from the apiserver side.

pippolo84 commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -1189,8 +1189,12 @@ data:
 
-  k8s-client-qps: {{ .Values.k8sClientRateLimit.qps | default $defaultK8sClientQPS | quote}}
-  k8s-client-burst: {{ .Values.k8sClientRateLimit.burst | default $defaultK8sClientBurst | quote }}
+{{- if .Values.k8sClientRateLimit.qps }}
```
Might be useful adding a comment here reporting what you said in PR description:

```
Configuring a custom rate limit via Helm will still set the rate limit for both Cilium Agent and Cilium Operator. If a separate custom rate limit is desired for the Operator, .Values.operator.extraArgs can be used to supply the flags directly to the deployment.
```

pippolo84 replied:
Alternatively, adding that to the OperatorK8sClientQPS{Limit,Burst} docstrings.
What is concerning me is the counterintuitive behavior of having two separate options with different defaults but the same knob to tweak them (although I understand the reason).

thorn3r replied:
I like your suggestion of moving everything under `operator/`. If we do that, we could use different names for the options and separate helm values. I think this makes the most sense. WDYT?

pippolo84 replied:
That would be nice!
In that case, we might want to add something in `Documentation/operations/upgrade.rst` to highlight that changing the operator's client QPS and Burst will require setting new options.

pippolo84 reviewed (commented):
Thanks!

From a functional point of view, changes LGTM, but I think a different way of organizing the code specific for the operator would be better.

squeed commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -1189,8 +1189,12 @@ data:
+{{- if .Values.k8sClientRateLimit.qps }}
+  k8s-client-qps: {{ .Values.k8sClientRateLimit.qps | quote}}
+{{- end }}
+{{- if .Values.k8sClientRateLimit.burst }}
```
nit: using `with` is a bit more idiomatic

squeed commented on operator/cmd/root.go:
```diff
@@ -98,11 +98,12 @@ var (
 		gops.Cell(defaults.GopsPortOperator),
 
-		// Provides Clientset, API for accessing Kubernetes objects.
-		k8sClient.Cell,
```
Suggestion: make the `k8sclient.Cell` and `ClientBuilderCell` take a `ClientParams` object, which is provided by an Agent or Operator cell.

squeed replied:
Looks like that's just what you did, looks nice :-)

thorn3r committed:
operator: allow configurable K8s client rate limit with higher defaults

The rate limit for K8s client(s) for Cilium Operator can now be
configured separately from that of Cilium Agent with the new
`--operator-k8s-client-qps` and `--operator-k8s-client-burst` flags.

To facilitate sharing a set of default configuration options for all
clients, the `Config` struct was split into `SharedConfig` and
`ClientParams`, the latter of which contains options expected to differ.

With K8s 1.26 reaching EOL, the default rate limit for Cilium Agent K8s
clients is now determined by the default config, rather than relying on
the upstream client library.

Since the Operator is a Deployment and does not scale linearly with
cluster size like the Agent DaemonSet, the rate limit can be increased
to improve performance at scale without heavily impacting control plane
resources. The default rate limit for Operator clients has been increased
to 100 QPS and 200 Burst.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

pippolo84 reviewed (approved):
Nice! :100:

qmonnet reviewed (approved):
Looks good from my side, thanks!
